src/core/namespace.c

```diff
@@ -687,7 +709,17 @@ static int mount_private_dev(MountEntry *m) {
         (void) symlink("/run/systemd/journal/dev-log", devlog);
 
         NULSTR_FOREACH(d, devnodes) {
-                r = clone_device_node(d, temporary_mount);
+                if (can_mknod) {
+                        r = clone_device_node(d, temporary_mount, can_mknod);
+                        if (r < 0) {
+                                if (r != -EPERM)
+                                        goto fail;
+
+                                can_mknod = false;
+                        }
+                }
+
+                r = clone_device_node(d, temporary_mount, false);
```

Comment:
If `can_mknod` is true, this will call `clone_device_node(..., true)` and then `clone_device_node(..., false)` for every node. Too ugly!

Let's step back a bit: on every call site, we do the same thing: call `clone_device_node(..., true)` if `can_mknod` is true, and then call `clone_device_node(..., false)`. This will be nicely simplified if the boolean argument is made a pointer, and `clone_device_node` sets it to false if it detects that we lack permissions.

Something like this:
```C
 NULSTR_FOREACH(d, devnodes) {
       r = clone_device_node(d, temporary_mount, &can_mknod);
       if (r < 0)
            goto fail;
}
```
and then rework `clone_device_node` to simply do the fallback on its own if `!*can_mknod`.